graph.cpp

```diff
@@ -49,27 +49,28 @@ bool Graph::addNode(const std::string& name, NodeBase* node) {
     }
 
     nodes_[name] = node;
+    node->setNameAndGraph(name, this);
     return true;
 }
 
-void Graph::removeNode(NodeBase* node) {
+void Graph::removeNode(const std::string& name) {
     ScopedLock lock(&mutex_);
 
-    std::map<string, NodeBase*>::iterator it = nodes_.find(node->name());
+    auto it = nodes_.find(name);
     if (it != nodes_.end()) {
         it->second->disconnectAllPins();
         it->second->disconnectAllStreams();
         nodes_.erase(it);
     }
 }
 
-NodeBase* Graph::getNodeByName(const std::string& name) {
+std::shared_ptr<NodeBase> Graph::getNodeByName(const std::string& name) {
     ScopedLock lock(&mutex_);
     return lockedGetNodeByName(name);
 }
 
-NodeBase* Graph::lockedGetNodeByName(const std::string& name) {
-    std::map<string, NodeBase*>::iterator it = nodes_.find(name);
+std::shared_ptr<NodeBase> Graph::lockedGetNodeByName(const std::string& name) {
+    auto it = nodes_.find(name);
```

Comment:
below we can return nullptr

Reply:
actually a default constructed shared_ptr..